Feat/incident hypothesis builder wip

AleksandrSekker opened the pull request.

saienkoo commented on src/icons/index.js:
```diff
@@ -371,5 +377,11 @@ export const Icons = {
 "account": clock,
 "name": cluster,
-"region": timezone
+"region": timezone,
```
можно использовать иконки из react-icons ?

saienkoo commented on tailwind.config.js:
```diff
@@ -50,7 +50,10 @@ module.exports = {
         "warm-blue": "#CCDAF8",
```
![image](https://user-images.githubusercontent.com/20907048/152002078-03f88b6e-6b14-474d-846c-c48eb43bff71.png)

иконка state в модалке и в аккордионе не совпадает

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisDetails/index.js:
```diff
@@ -99,52 +108,48 @@ export function HypothesisDetails({ nodePath, tree, setTree, api, ...rest }) {
           />
         </div>
       </div>
+      <div className="absolute top-8 right-20">
```
это вообще не хорошо. доделай компонент modal что бы в него можно было передавать соседей кнопки для закрытия. либо (если возможно) сделай так, что бы модалка вообще не рендерила кнопку, а обе кнопки ты передавал вместе со всем остальным как чилдренов. в любом случае сейчас я вижу что при нажатии на три точки есть эффект, а при нажатии на крестик нету - не хорошо

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/HypothesisNodeItem.js:
```diff
@@ -0,0 +1,59 @@
+      </span>
+    </div>
+    <div className="flex items-center h-12">
+      <Icon name={iconRightFirst} className="ml-3.5 mr-2.5" />
```
а если нужно будет 5 иконок? а если третью из пяти не нужно будет рендерить, что делать? это компонент с вполне себе конкретным назначением, можно попробовать сделать через пропсы (showMessageIcon, showSignalIcon) или дать больше контроля и позволить передавать нужные иконки сколько нужно и в нужном порядке

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/HypothesisNodeItem.js:
```diff
@@ -0,0 +1,59 @@
+    <div className="flex items-center h-12">
+      {renderChevronIcon()}
+      <div className="ml-2 bg-lighter-gray rounded-full p-2">
+        <Icon name={iconLeft} className="" />
```
react-icons

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/HypothesisNodeItem.js:
```diff
@@ -0,0 +1,59 @@
+        <Icon name={iconLeft} className="" />
+      </div>
+
+      <span
```
если здесь будет строка из трех символов, область для клика будет слишком маленькая

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/index.js:
```diff
@@ -0,0 +1,112 @@
+    setSelectedNodePath([...parentArray, node.id]);
+    setModalIsOpen(true);
+  };
+  const getStatusIcon = (status) => {
```
посмотри как я сделал для таблицы инцидентов
1. для статусов - enums
2. для каких-либо соответствий - маппинги с фолбеками

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/index.js:
```diff
@@ -0,0 +1,112 @@
+    <div>
+      <div className="flex align-baseline text-base font-semibold mb-6">
+        <h2 className="text-dark-gray mr-3 font-inter text-2xl">Action plan</h2>
+        <div className="bg-dark-blue rounded-full p-2">
```
на дизайне там очевидно кнопка, у которой должны быть соответствующие стили

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/index.js:
```diff
@@ -0,0 +1,112 @@
+                  {node.children.map((item, index) => (
+                    <div key={item.id} className={cx(index !== 0 && "mt-5")}>
+                      <div className="flex align-baseline text-sm font-medium mb-2.5">
+                        <p className="text-dark-gray mr-2.5 mt-0.5 font-medium">
```
```suggestion
                        <span className="text-dark-gray mr-2.5 mt-0.5 font-medium">
```

это не параграф, а просто надпись которая идет в одной строке с кнопкой.

saienkoo replied:
я думаю для одного инцидента в контенте аккордиона эта надпись должна появляться один раз, у тебя - для каждой гипотезы

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/index.js:
```diff
@@ -0,0 +1,112 @@
+                      <div className="ml-7">
+                        <div>
+                          <div className="flex my-4 align-baseline text-sm font-medium mb-2.5">
+                            <p className="text-dark-gray mr-2.5 mt-0.5 font-medium">
```
```suggestion
                            <span className="text-dark-gray mr-2.5 mt-0.5 font-medium">
```

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/index.js:
```diff
@@ -0,0 +1,112 @@
+                    onOpenModal={handleOpenModal}
+                  />
+                </Disclosure.Button>
+                <Disclosure.Panel className="bg-light-blue p-5  mt-4 ml-1 rounded-8px border border-dashed">
```
![image](https://user-images.githubusercontent.com/20907048/152016600-63a3540d-a244-4878-b593-1d3ce3b0b230.png)

элементы не на одной линии

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisNodeNew/index.js:
```diff
@@ -0,0 +1,112 @@
+                          <Icon name="addButton" className="w-2.5 h-2.5" />
+                        </div>
+                      </div>
+                      <HypothesisNodeItem
+                        iconLeft={getStatusIcon(item.status)}
+                        text={item.title || "(none)"}
+                        iconRightFirst="message"
+                        iconRightSecond="diagram"
+                        image="https://images.unsplash.com/photo-1531427186611-ecfd6d936c79?ixlib=rb-=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=8&w=256&h=256&q=80"
+                      />
+                      <div className="ml-7">
+                        <div>
+                          <div className="flex my-4 align-baseline text-sm font-medium mb-2.5">
+                            <p className="text-dark-gray mr-2.5 mt-0.5 font-medium">
+                              Pottential Solution
+                            </p>
+                            <div className="bg-dark-blue rounded-full p-2">
+                              <Icon name="addButton" className="w-2.5 h-2.5" />
+                            </div>
+                          </div>
+                          {item.children.map(({ id, title, status }) => (
+                            <div key={id} className="flex flex-col mb-0.5">
+                              <HypothesisNodeItem
+                                iconLeft={getStatusIcon(status)}
+                                text={title || "(none)"}
+                                iconRightFirst="message"
+                                image="https://images.unsplash.com/photo-1531427186611-ecfd6d936c79?ixlib=rb-=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=8&w=256&h=256&q=80"
+                              />
+                            </div>
+                          ))}
+                        </div>
+                      </div>
```
это можно как-то вынести в отдельный компонент или может даже несколько? только если ты их понятно назовешь.

saienkoo committed:
feat: hypothesis builder in process

feat: incident and hypothesis modal

feat: update layout for incident and hypothesis

feat: update hypothesis builder

feat: update icons showing

fix: make default avatars when image broken

fix: update default avatar hypothesis row

fix: remove accordion, rollback icon, change modal

fix: remove dropdown menu, replace cx to slsx

fix: move default avatar icon to image component remove unneded icon

refactoring: move message and chart icon to react icons

saienkoo committed:
fix: topology builder improvements

saienkoo commented on src/components/HypothesisBuilder/components/EvidenceSection/index.js:
```diff
@@ -1,22 +1,35 @@
 
 export function EvidenceItem({ evidence }) {
   if (evidence.type === "log") {
     return <LogsTable logs={evidence.evidence} title="" />;
```
![image](https://user-images.githubusercontent.com/20907048/153417878-b22a8f4b-4567-4ab2-9509-ebadaeff1d10.png)

saienkoo commented on src/components/HypothesisBuilder/components/EvidenceSection/index.js:
```diff
@@ -1,22 +1,35 @@
   }
   if (evidence.type === "topology") {
-    return <TopologyCard topologyId={evidence.evidence.id} />
+    return <TopologyCard topologyId={evidence.evidence.id} />;
```
карточка маленькая

OleksandrSekker committed:
feat: update hypothesis detail modal

saienkoo commented on tailwind.config.js:
```diff
@@ -53,7 +53,9 @@ module.exports = {
         "darker-gray": "#9CA3AF",
         "lighter-gray": "#F3F4F6",
         "dark-gray": "#242423",
+        "gray-dark": "#383A40",
```
подбери похожий цвет из доступных в tailwind. это уже перебор, что у нас dark-gray и тут же за ним gray-dark

saienkoo commented on src/components/HypothesisBuilder/index.js:
```diff
@@ -50,11 +50,13 @@ export function HypothesisBuilder({
       <Modal
         open={modalIsOpen}
-        onClose={() => setModalIsOpen(false)}
-        cardClass="w-full overflow-y-scroll"
```
ты уверен, что без `overflow-y-scroll` будет правильно работать? может лучше overflow-y-auto или как там он ?

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisRootChildrenList/index.js:
```diff
@@ -3,7 +3,7 @@ import clsx from "clsx";
 import { HypothesisBar } from "../HypothesisBar";
 
-export const HypothesisRootChildrenList = ({ node }) => (
+export const HypothesisRootChildrenList = ({ node, onTitleClick }) => (
```
лучше подели на два свойста - onIssueClick, onSolutionClick

saienkoo commented on src/components/HypothesisBuilder/components/HypothesisDetails/index.js:
```diff
@@ -186,7 +176,10 @@ export function HypothesisDetails({ nodePath, tree, setTree, api, ...rest }) {
   return (
-    <div className={`text-md font-medium text-gray-800 ${className}`} {...rest}>
+    <div
+      className={`text-lg font-medium text-gray-dark font-semibold ${className}`}
```
почему не clsx ?

saienkoo commented on src/components/HypothesisBuilder/components/EvidenceSection/index.js:
```diff
@@ -1,34 +1,17 @@
 
 export function EvidenceItem({ evidence }) {
   if (evidence.type === "log") {
-    return <LogsTable logs={evidence.evidence} title="" />;
```
`return <LogsTable logs={evidence.evidence} title="" />;`

saienkoo commented on src/components/HypothesisBuilder/components/EvidenceSection/index.js:
```diff
@@ -1,34 +1,17 @@
-    return <TopologyCard topologyId={evidence.evidence.id} />;
+    return <TopologyCard topologyId={evidence.evidence.id} size="small" />;
   }
-  return (
```
`return null`

saienkoo commented on src/components/HypothesisBuilder/components/CommentsSection/index.js:
```diff
@@ -76,15 +75,15 @@ export function CommentsSection({
                         </span>
                       </div>
-                      <p className="mt-0.5 text-xs text-gray-400">
+                      <p className="mt-0.5 text-medium-gray text-sm leading-5 font-normal">
```
новые цвета старайся не добавлять, если без них не обойтись, и пытайся подобрать из tailwindcss(но и тут тоже в крайности не бросайся). часто в фигре прямо так и написано - gray/900 например.

saienkoo commented on src/components/HypothesisBuilder/components/CommentsSection/index.js:
```diff
@@ -26,14 +26,13 @@ export function CommentsSection({
         <textarea
           disabled={isLoading}
-          className="w-full border-gray-200 resize-none rounded-6px text-base leading-6 font-normal font-inter outline-none text-dark-gray mb-2"
+          className="w-full border-gray-200 resize-none rounded-6px text-base leading-6 font-normal font-inter outline-none placeholder-darker-gray mb-2"
```
это text-gray-400. пересмотри все свои цвета здесь

saienkoo commented on src/components/EvidenceLogTable/EvidenceLogTable.js:
```diff
@@ -0,0 +1,21 @@
+import { BsFillBarChartFill } from "react-icons/all";
+import dayjs from "dayjs";
+
+export const EvidenceLogTable = ({ evidence }) => (
```
так то это не таблица получилась, хотя в названии компонента table. вообще, если я правильно понял, в конце этот компонент не понадобится.

saienkoo committed:
fix: more fluid approach to render tree, open details modal on click on any hypothesis

OleksandrSekker committed:
fix: make fixes

saienkoo commented on tailwind.config.js:
```diff
@@ -50,17 +50,22 @@ module.exports = {
-        "light-green": "#98C373"
+        "light-green": "#98C373",
+        "bright-green": "#459E45",
+        "warm-green": "#AAA526",
```
куча новых цветов. все они нужны?

AleksandrSekker replied:
<img width="370" alt="Screenshot 2022-02-15 at 1 48 02 PM" src="https://user-images.githubusercontent.com/32077311/154056536-1b6e55e7-3b76-440b-80ae-ba7810c5164a.png">
они используются в компоненте HypothesisBar

saienkoo committed:
fix: handle cases when node's title is empty or space char

saienkoo committed:
feat: start implementing react-query on incident table

saienkoo committed:
feat: use react-query mutation to update hypothesis

OleksandrSekker committed:
feat: update hypothesis create modal, make fixes

saienkoo commented on src/components/Avatar/index.js:
```diff
@@ -72,7 +71,7 @@ export const Avatar = ({
 
 Avatar.propTypes = {
   size: PropTypes.string,
-  srcList: PropTypes.oneOf([
```
ты уверен, что теперь нету ошибки? по-моему оно ругалось что передается null

AleksandrSekker replied:
да, сейчас в консоле нет этой ошибки, оно ругалось из того что oneOf принимает значения,  а внутри типы данных
<img width="876" alt="Screenshot 2022-02-16 at 2 32 56 PM" src="https://user-images.githubusercontent.com/32077311/154265354-6945ce8e-1646-4bdc-9fc2-5327854ed8c8.png">

saienkoo commented on src/components/Change/index.js:
```diff
@@ -85,14 +86,12 @@ function classNames(...classes) {
       <ul role="list" className="-mb-4">
         {timeline.map((event, eventIdx) => (
-          <li key={event.id}>
+          <li key={uuidv4()}>
```
не нужно такое делать, вычисления таких хешей могут быть ресурсозатратными. лучше укажи eventIdx и заглуши варнинг от еслинта. и опять таки - тебе бы разобраться зачем именно указывается key. твой вариант еще хуже, чем использовать индекс мапа

AleksandrSekker replied:
не совсем понял что ты имеешь ввиду под зашлушить ворнинг, отключить это правило в этой строке или как?

saienkoo replied:
да, если ты в key передашь eventIdx и eslint начнет ругаться - заглуши варнинг только для этой строки

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -0,0 +1,122 @@
+  onHypothesisCreated
+}) => {
+  const [commentTextValue, setCommentTextValue] = useState("");
+  const [nameValue, setNameValue] = useState("");
```
Это что? У тебя есть пекедж react-hook-form, который позволяет тебе строить формы используя его логику. Если с комментарием еще ладно - по идее он будет отправляться отдельно, то имя должно быть рядом со статусом в форме

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -0,0 +1,122 @@
+
+  const watchStatus = watch("status");
+  const onSubmit = () => {
+    const newNodeID = handleAddNode(nodePath);
```
эта функция вставляет в дерево новый стандартный итем и возвращает тебе id. используя react-query, мы хотим отойти от этого. что бы потом данные шли в одну сторону - данные получили, обработали и они пошли вниз по дереву компонентов. если делать как было у них - каждый дочерний компонент получается сам может редактировать всё дерево, и должен знать как это правильно сделать что бы ничего не сломалось

saienkoo replied:
```suggestion
    const newNodeID = uuidv4();
```

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -0,0 +1,122 @@
+        }
+      });
+    }
+    onHypothesisCreated();
```
модалка закроется раньше, чем гипотеза будет создана.

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -0,0 +1,122 @@
+  });
+
+  const watchStatus = watch("status");
+  const onSubmit = () => {
```
```suggestion
  const onSubmit = async () => {
```

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -0,0 +1,122 @@
+  const onSubmit = () => {
+    const newNodeID = handleAddNode(nodePath);
+    if (api?.createMutation) {
+      api.createMutation.mutate({
```
```suggestion
      await api.createMutation.mutateAsync({
```

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -0,0 +1,122 @@
+  const { handleAddNode } = treeFunctions;
+  const user = useUser();
+  const node = getNode(tree, nodePath);
+  const statusItems = {
```
Это обязательно должно быть внутри компонента и пересчитываться на каждый рендер?

AleksandrSekker replied:
treeFunctions уже не нужен,  хук useUser не может вызывать вне компонента (eslint ворнинг), а getNode использует данные с пропсов

saienkoo replied:
Мой комментарий стоит только на одной строчке - обьявлении statusItems обьекта. в src/components/HypothesisBuilder/components/HypothesisDetails/index.js, откуда ты его похоже скопировал, он обьявлен и посчитан вне компонента

saienkoo commented on src/components/HypothesisBuilder/index.js:
```diff
@@ -70,7 +81,48 @@ export function HypothesisBuilder({
+            >
+              {(heirarchyProps) => (
+                <>
+                  <CreateHypothesis
```
В итоге тебе не нужна будет handleAddNode и можно не использовать NestedHeirarchy в этом случае

saienkoo commented on src/components/HypothesisBuilder/index.js:
```diff
@@ -38,12 +41,20 @@ export function HypothesisBuilder({
-              defaultEditMode={defaultEditMode}
-              api={api}
-            />
+            {(heirarchyProps) => (
```
я тебе вчера пытался обьяснить что если передавать сюда чилдренов в виде рендер-фунгкции, то можно получить сформированные пропсы, рендерить фрагмент и в него всё, что нужно. тоесть не рендерить NestedHeirarchy больше одного раза

saienkoo commented on src/pages/incident/incident-details.js:
```diff
@@ -81,12 +85,14 @@ export function IncidentDetailsPage() {
       queryClient.invalidateQueries(createIncidentQueryKey(incidentId));
     }
   });
+  const createMutation = useCreateHypothesisMutation({
```
посмотри что я скоро запушу - можно попытаться организовать optimistic updates. тоесть мутация добавляет новые данные в кеш до отправки запроса и если была ошибка - отменяет это. но делай это в самую последнюю очередь, и уточни у меня еще дополнительно перед этим

saienkoo commented on src/pages/incident/incident-details.js:
```diff
@@ -81,12 +85,14 @@ export function IncidentDetailsPage() {
     }
   });
+  const createMutation = useCreateHypothesisMutation({
+    onSettled: () => {
```
я перенес эти колбеки в саму мутацию. посмотри

saienkoo committed:
feat: react-query improvements and using optimistic updates to keep instant-save behavior from v2

OleksandrSekker committed:
feat: makes updates and fixes

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -35,31 +34,40 @@ export const CreateHypothesis = ({
         incidentId: api.incidentId,
         params: {
-          title: nameValue,
+          title: getValues("title"),
```
Почему в соседних строках разный подход? Либо везде getValues, либо везде значения с watch. Но надежнее будет getValues, поскольку он возвращает из стейта хука useForm актуальные данные. А значения с watch могут быть устаревшими, так как чисто теоретически нету гарантий, что между очередным изменением инпута и сабмитом формы обязательно будет хотя бы один рендер

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -35,31 +34,40 @@ export const CreateHypothesis = ({
-          title: nameValue,
+          title: getValues("title"),
           type: node.type,
           status: watchStatus
```
Судя по пот этой документации https://react-hook-form.com/api/useform/getvalues
возможно получится хранить данные в под-секциях и удобно получать их

стоит попытаться такое сделать 5-10 минут
```
  const { control, watch, getValues, setValue, handleSubmit } = useForm({
    defaultValues: {
      hypothesis: {
        status: node.status || Object.values(statusItems)[2].value,
        title: ""
      },
      evidence: {},
      comment: {
        text: ""
      }
    }
  });

const hypothesisValues = getValues("hypothesis");
      const evidenceValues = getValues("evidence");
      const commentValues = getValues("comment");

        <input
          id="name"
          type="text"
          placeholder="Type name"
          onChange={(e) => {
            setValue("hypothesis.title", e.target.value);
          }}
          value={nameValue}
          className="..."
        />
        <Dropdown
          control={control}
          name="hypothesis.status"
          className="mb-4 w-72"
          items={statusItems}
          label="State"
        />
```

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -35,31 +34,40 @@ export const CreateHypothesis = ({
         }
       });
     }
+    await handleComment(getValues("comments"));
```
Если if (api?.createMutation) не выполняется, то мы всёравно создаем комментарий? И к чему именно он создастся, если даже в случае успешного создания гипотезы, ты не пытаешься получить ее и как-то дальше использовать?

saienkoo commented on src/components/HypothesisBuilder/components/CreateHypothesis/index.js:
```diff
@@ -101,22 +111,20 @@ export const CreateHypothesis = ({
         </h3>
         <textarea
           className="w-full border-gray-200 resize-none rounded-6px text-base leading-6 font-normal font-inter outline-none placeholder-gray-400"
-          onChange={(e) => setCommentTextValue(e.target.value)}
```
это обычный инпут без заморочек. посмотри в документации react-hook-form как можно компактнее нацепить на него логику

OleksandrSekker committed:
fix: make fixes in create hypothesis

OleksandrSekker committed:
feat: Details components

saienkoo committed:
fix: use correct text color class

saienkoo commented on src/components/Details/index.js:
```diff
@@ -0,0 +1,257 @@
+      "https://images.unsplash.com/photo-1520785643438-5bf77931f493?ixlib=rb-=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=8&w=256&h=256&q=80"
+  }
+];
+export const Details = () => {
```
Что за Details? Абсолютно что это за компонент, как его использовать.

saienkoo commented on src/components/Details/index.js:
```diff
@@ -0,0 +1,257 @@
+import { Avatar } from "../Avatar";
+import { IncidentPriority } from "../../constants/incident-priority";
+
+const SingleValue = ({ children, ...props }) => (
```
Что за SingleValue? За что он отвечает? как его использовать? название должно описывать смысл компонент + в этом случае суффикс SingleValue
`IncidentPrioritySingleValue`

весь этот поток компонентов можно вынести а отдельный файл где-то рядом, например `select-components.js`

saienkoo commented on src/components/Details/RowDetail.js:
```diff
@@ -0,0 +1,15 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+export const RowDetail = ({ title, value }) => (
```
при переименовании Details это тоже нужно будет назвать получше

saienkoo commented on src/components/Details/index.js:
```diff
@@ -0,0 +1,257 @@
+];
+export const personRespondents = [
+  {
+    label: "Lindsay Walton",
```
в данных инцидента, которые возвращает апи, есть respondents, commander?

saienkoo commented on src/components/HypothesisBuilder/data.js:
```diff
@@ -49,7 +49,8 @@ export const hypothesisInitialFields = {
   links: [],
-  comments: []
+  comments: [],
+  created_by: []
```
это обьект должен быть

saienkoo commented on src/index.css:
```diff
@@ -101,5 +119,37 @@
+    @apply h-btn-round-lg w-btn-round-lg text-xs;
+  }
+
+  .btn-round-xl {
```
на дизайне тоже 6 размеров?

AleksandrSekker replied:
<img width="616" alt="Screenshot 2022-02-22 at 10 58 19 AM" src="https://user-images.githubusercontent.com/32077311/155097924-b5047dbf-8428-4203-9589-825ad0e18ff8.png">
5 размеров, там стили для кнопки, для disabled кнопки и их размеры
<img width="1127" alt="Screenshot 2022-02-22 at 11 02 12 AM" src="https://user-images.githubusercontent.com/32077311/155098525-26df6e4d-a28d-483d-b347-c1fbfc106ce1.png">

saienkoo commented on src/pages/incident/incident-details.js:
```diff
@@ -189,6 +186,7 @@ export function IncidentDetailsPage() {
               <Changelog />
             </section>
           </div>
+          <Details />
```
на дизайне эта форма в другом месте. почему она здесь?

AleksandrSekker replied:
<img width="342" alt="Screenshot 2022-02-22 at 11 06 14 AM" src="https://user-images.githubusercontent.com/32077311/155099508-83ae7c1d-b398-4ac7-873f-b89aee00414a.png">
то есть эта форма должна быть вместо этой или как?

saienkoo replied:
получается вместо этой, да

saienkoo commented on src/components/Details/index.js:
```diff
@@ -0,0 +1,257 @@
+        title="Respondents"
+        value={
+          <Select
+            name="respondents"
```
я не могу удалить выбранные итемы в этом селекте

AleksandrSekker replied:
когда селект выбраный итемы можно удалять как обычный текст, нажав backspace на клавиатуре. А как иначе это должно работать? в плане что на дизайне нет крестиков возле итемов или чего то похожего на удаление

saienkoo replied:
backspace не подойдет, если у тебя 10 выбранных юзеров, а удалить нужно первого. в react-select можно передать свойство, что бы он не удалял выбранные итемы из меню

saienkoo commented on src/components/Details/index.js:
```diff
@@ -0,0 +1,257 @@
+      <RowDetail
+        title="Priority"
+        value={
+          <Select
```
не думаю, что в этом селекте стоит позволять текст в инпут для поиска

saienkoo commented on src/components/Details/index.js:
```diff
@@ -0,0 +1,257 @@
+
+export const personCommanders = [
+  {
+    label: "Lindsay Walton",
```
коммандера ты сейчас уже вполне можешь получить, если дополнить запрос вот так
```
incident?id=eq.597806a0-4752-4a2f-87a4-bc5fff793f2b&select=*,commander_id(name,avatar),hypothesis!hypothesis_incident_id_fkey(*,created_by(name,avatar))
```

насчет респондентов не понятно, а обновление статуса и коммандера можно реализовать. делай это в последнюю очередь, когда всё остальное починишь

OleksandrSekker committed:
feat: update incedent details component

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,175 @@
+  updateStatusHandler,
+  textButton
+}) => {
+  const commandersArray = useMemo(
```
добавь комментарий
// Temporary mock, in the future you need to replace it with an array of real users received from the api

saienkoo commented on src/components/IncidentDetails/data.js:
```diff
@@ -0,0 +1,20 @@
+export const personRespondents = [
```
добавь комментарий что это временные данные, пока не будет получаться список потенциальных респондеров с апи или откуда-то еще

saienkoo commented on src/components/IncidentDetails/data.js:
```diff
@@ -0,0 +1,20 @@
+export const personRespondents = [
+  {
+    label: "Lindsay Walton",
+    value: "Lindsay Walton",
```
в value должны быть id, даже в dummy данных просто для наглядности

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,175 @@
+  const { control, getValues, watch } = useForm({
+    defaultValues: {
+      tracking: "123456",
+      started: dayjs(incident.created_at).format("DD.MM.YYYY") || "01.01.2022",
```
```suggestion
      started: incident.created_at,
```

Слабо представляю себе сценарий, когда у полученного с апи инцидента не будет указано created_at значение. даже если там будет undefined - dayjs вернет текущее время и ты его отформируешь. а значит

1. твой фолбек на 01.01.2022 не имеет смысла
2. странный фолбек сам по себе. лучше уж "прочерк" указывать. но не в логике формы. в логике формы должны быть адекватные значения, а форматирование происходит уже во времея рендера в каждом конкретном случае

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,175 @@
+        title="Started"
+        value={
+          <span className="text-dark-gray text-sm font-normal">
+            {getValues("started")}
```
```
const formatDate = (date, fallback = '-') => date ? dayjs(date).format(...) : fallback;
```

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,175 @@
+    defaultValues: {
+      tracking: "123456",
+      started: dayjs(incident.created_at).format("DD.MM.YYYY") || "01.01.2022",
+      duration: dayjs(incident.created_at).fromNow() || "13 days",
```
фолбек непонятно о чем. мы тут уже с реальными данными работаем, а не просто верстаем. просто убери это, а уже в jsx можешь считать нужно значение

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,175 @@
+        }
+      />
+      <IncidentDetailsRow
+        title="Commanders"
```
на дизайне и Commanders, и Respondents написаны во множественном числе. но у инцидента коммандер указан и он один, а значит сделай наверное что бы этот инпут можно было очистить - была та кнопочка-крестик. а дальше, имея два рабочих примера, кто-то доделает как нужно

saienkoo commented on src/components/IncidentDetails/select-components.js:
```diff
@@ -0,0 +1,82 @@
+    </div>
+  </components.Option>
+);
+export const priorities = [
```
судя по названию в этом файле должны находиться только компоненты для селекта. перенеси в сам компонент или в data.js

OleksandrSekker committed:
fix: make fixes

saienkoo commented on src/components/IncidentDetails/data.js:
```diff
@@ -0,0 +1,52 @@
+// Temporary mock, in the future you need to replace it with an array of real respondents received from the apis
```
так и задумано, что комментарий тут, а не над personRespondents ?

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,181 @@
+    }
+  });
+  watch();
+  const formatDate = (date, fallback = "-") => {
```
зачем так сложно? если хочешь - можно где-то за пределами компонента создать обьект formatDateUtils и в нем две функции - toDate, toDucation. это должны быть чистые функции. дальше или просто используй их, или если результат их выполнения используется больше чем 1 раз - в useMemo засунь

saienkoo replied:
почему ты уже который раз пытаешься сделать именно функцию, которая вызывается и что-то из себя возвращает, и потом вызываешь ее 2+ раз? а что, если ее прийдется потом вызвать 1000 раз в один рендер? что будет с производительностью?

saienkoo replied:
в общем - почитай что такое чистые функции, старайся функции создавать чистыми или хотя бы почище при возможности, так как их проще понять, тестировать и использовать.

ты должен был создать в каком-либо виде две такие функции и использовать:

```
const DATE_FORMAT = "DD.MM.YYYY";

const formatDate = ({ date, format = DATE_FORMAT, fallback = "-" }) =>
  date ? dayjs(date).format(format) : fallback;
const dateToDuration = ({ date, withoutSuffix = false, fallback = "-" }) =>
  date ? dayjs(date).fromNow(withoutSuffix) : fallback;
```

либо написать примерно такие useMemo()

```
  const watchCreatedAt = watch("created_at");

  const formattedCreatedAt = useMemo(
    () => (watchCreatedAt ? dayjs(watchCreatedAt).format(DATE_FORMAT) : "-"),
    [watchCreatedAt]
  );
  const formattedDuration = useMemo(
    () => (watchCreatedAt ? dayjs(watchCreatedAt).fromNow() : "-"),
    [watchCreatedAt]
  );
```

либо скомбинировать эти подходы

```
  const watchCreatedAt = watch("created_at");

  const formattedCreatedAt = useMemo(
    () => formatDate({ date: watchCreatedAt }),
    [watchCreatedAt]
  );
  const formattedDuration = useMemo(
    () => dateToDuration({ date: watchCreatedAt }),
    [watchCreatedAt]
  );
```

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,181 @@
+  });
+  watch();
+  const formatDate = (date, fallback = "-") => {
+    const getDate = date ? dayjs(date).format("DD.MM.YYYY") : fallback;
```
почему название переменной getDate, если это не функция?

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,181 @@
+  watch();
+  const formatDate = (date, fallback = "-") => {
+    const getDate = date ? dayjs(date).format("DD.MM.YYYY") : fallback;
+    const getDuration = date ? dayjs(incident.created_at).fromNow() : fallback;
```
почему название getDuration, если это не функция?
почему ты проверяешь date из аргументов, а в dayjs передаешь incident.created_at из внешнего скоупа?

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,181 @@
+    defaultValues: {
+      tracking: "123456",
+      started: incident.created_at,
+      duration: incident.created_at,
```
```suggestion
      created_at: incident.created_at,
```

здесь должны быть данные инцидента. что там с них ты уже отрендеришь - другой вопрос. хоть и 10 надписей с одной created_at даты

OleksandrSekker committed:
fix: make fixes in incedent details

saienkoo commented on src/components/HypothesisBuilder/components/EvidenceSection/index.js:
```diff
@@ -36,13 +39,20 @@ export function EvidenceSection({
-          <div className="py-2 px-4 text-sm text-gray-400">No evidence</div>
+          <div className="border border-dashed border-gray-300 rounded-8px flex justify-center pt-8 mt-2.5">
+            <div className="flex flex-col mx-auto mb-6">
+              <FiDatabase className="h-10 w-10 text-gray-400 mx-auto" />
```
попробуй с фигмы экспортировать svg нужной иконки и сделать из этого компонент в src/omponents/src-icons

https://www.sanity.io/guides/import-svg-files-in-react#c125367dce7c

OleksandrSekker committed:
feat: update database plus icon

OleksandrSekker committed:
fix: change margins in incident details

OleksandrSekker committed:
fix: fix svg-icons folder name

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,203 @@
+          </a>
+        }
+      />
+      <div className="-mt-1.5">
```
и какой в этом смысл? сделали статичные нижние отступы в виде паддингов, потом добавляя лишнюю разметку отрицательными паддингами сдвигаем элементы. ты представляешь, какая морока потом может быть тебе же что-то менять в этом через месяц?

не нужно вообще таким элементам делать свои отступы, потому что если понадобится отрендерить такую строку где-то в компактном месте - твой же pb-4 будет тебе мешать

saienkoo commented on src/components/IncidentDetails/index.js:
```diff
@@ -0,0 +1,203 @@
+      />
+      <div className="-mt-1.5">
+        <IncidentDetailsRow
+          title="Status Page"
+          value={
```
```suggestion
          title="Status Page"
          className="mb-2.5"
          value={
```

saienkoo commented on src/pages/incident/incident-details.js:
```diff
@@ -146,44 +142,53 @@ export function IncidentDetailsPage() {
+              status === "open" ? "Mark as resolved" : "Mark as reopen"
+            }
+          />
+          {/* <div className="bg-white px-4 py-5  shadow sm:rounded-lg sm:px-6"> */}
```
удали закомментированное

OleksandrSekker committed:
feat: update database plus icon

OleksandrSekker committed:
fix: delete comments
